core/modules/qhttp/Server.cc

```diff
@@ -0,0 +1,198 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "Server.h"
+
+// System headers
+#include <memory>
+#include <string>
+
+// Third-party headers
+#include "boost/asio.hpp"
+#include "boost/asio/steady_timer.hpp"
+#include "boost/regex.hpp"
+
+// Local headers
+#include "qhttp/AjaxEndpoint.h"
+#include "qhttp/StaticContent.h"
+
+namespace asio = boost::asio;
+namespace ip = boost::asio::ip;
+
+namespace lsst {
+namespace qserv {
+namespace qhttp {
+
+
+Server::Ptr Server::create(asio::io_service& io_service, unsigned short port)
+{
+    return std::shared_ptr<Server>(new Server(io_service, port));
+}
+
+
+unsigned short Server::getPort()
+{
+    return _acceptor.local_endpoint().port();
+}
+
+
+Server::Server(asio::io_service& io_service, unsigned short port)
+:
+    _io_service(io_service),
+    _acceptor(io_service, ip::tcp::endpoint(ip::tcp::v4(), port))
+{
+}
+
+
+void Server::addHandler(std::string const& method, std::string const& pattern, Handler handler)
+{
+    auto &handlers = _pathHandlersByMethod[method];
+    handlers.resize(handlers.size() + 1);
+    auto &phandler = handlers.back();
+    phandler.path.parse(pattern);
+    phandler.handler = handler;
+}
+
+
+void Server::addHandlers(std::initializer_list<HandlerSpec> const& handlers)
+{
+    for(auto& handler: handlers) {
+        addHandler(handler.method, handler.pattern, handler.handler);
+    }
+}
+
+
+void Server::addStaticContent(std::string const& pattern, std::string const& rootDirectory)
+{
+    StaticContent::add(*this, pattern, rootDirectory);
+}
+
+
+AjaxEndpoint::Ptr Server::addAjaxEndpoint(const std::string& pattern)
+{
+    return AjaxEndpoint::add(*this, pattern);
+}
+
+
+void Server::accept()
+{
+    auto socket = std::make_shared<ip::tcp::socket>(_io_service);
+    _acceptor.async_accept(
+        *socket,
+        [this, socket](boost::system::error_code const& ec) {
+            accept(); // start accept for the next incoming connection
+            if (!ec) {
+                ip::tcp::no_delay option(true);
+                socket->set_option(option);
+                _readRequest(socket);
+            }
+        }
+    );
+}
+
+
+void Server::_readRequest(std::shared_ptr<ip::tcp::socket> socket)
+{
+    auto timer = std::make_shared<asio::steady_timer>(_io_service);
+    timer->expires_from_now(std::chrono::seconds(300));
```

Comment:
300 seconds is default timeout for HTTP requests in Apache since 1.2, Mozilla Firefox browsers, and many other Web applications.
I would argue this value should be configurable either via the constructor of the class, or via some other mechanism.